src/coreclr/jit/valuenum.cpp

```diff
@@ -8612,10 +8626,19 @@ bool Compiler::fgValueNumberConstLoad(GenTreeIndir* tree)
     FieldSeq* fieldSeq = nullptr;
     if (varTypeIsIntegral(tree) && GetStaticFieldSeqAndAddress(vnStore, tree->gtGetOp1(), &address, &fieldSeq))
     {
-        assert(fieldSeq->GetKind() == FieldSeq::FieldKind::SimpleStaticKnownAddress);
         CORINFO_FIELD_HANDLE fieldHandle = fieldSeq->GetFieldHandle();
+        ssize_t              byteOffset  = 0;
+        if (fieldSeq->GetKind() == FieldSeq::FieldKind::SimpleStaticKnownAddress)
+        {
+            byteOffset = address - fieldSeq->GetOffset();
+        }
+        else
+        {
+            assert(address == -1);
+            assert(fieldSeq->GetKind() == FieldSeq::FieldKind::SimpleStatic);
+            byteOffset = fieldSeq->GetOffset();
```

Comment:
This is still not quite right. `byteOffset` is meant to be the offset of data at which we're looking relative to the start of the (static) field. `fieldSeq->GetOffset()` provides the offset relative to the static base.

The right formula for `byteOffset` in:
```
ADD(<ICON_STATIC>, OFFSET FldSeq)
```
Would be `OFFSET - FldSeq->GetOffset()`, which is why I would like to echo my earlier suggestion of replacing the out parameter of `address` with `byteOffset` in `GetStaticFieldSeqAndAddress`. The caller only needs the latter.

Reply:
@SingleAccretion I thought FldSeq->GetOffset() == OFFSET value in case of GTF_ICON_FIELD_SEQ 🤔

Reply:
The field sequence is obtained here from the IR:
https://github.com/dotnet/runtime/blob/4649e436e955922fabdfc15cfb9df7b116d30452/src/coreclr/jit/valuenum.cpp#L8551-L8562

The `FldSeq->GetOffset() != OFFSET` case arises, for example, when we have a struct field added to the sequence and morph folds `ADD(ADD(<handle>, STATIC_FIELD_OFFSET), STRUCT_FIELD_OFFSET)` into `ADD(<handle>, STATIC_FIELD_OFFSET + STRUCT_FIELD_OFFSET)`.